### PR TITLE
[Oracle] Implement the save to project

### PR DESCRIPTION
**Pull request by troopa81**
This PR implements project saving to the Oracle database.

### CODE REVIEW

**Comment by troopa81** [comment]
@elpaso or @rldhont Do you mind reviewing this PR?

**Comment by nyalldawson** [comment]
@elpaso do you feel confident reviewing this one as part of the 3.26 funded freeze work?

**Comment by elpaso** [comment]
> @elpaso do you feel confident reviewing this one as part of the 3.26 funded freeze work?

Sure.

**Review by elpaso — APPROVED** [review]
Nice work!

I just left some minor comments.

File: src/providers/oracle/qgsoracleprojectstorage.cpp [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+      QDateTime lastModifiedUtc = QDateTime::fromString( lastModifiedTimeStr, Qt::ISODate );
+      lastModifiedUtc.setTimeSpec( Qt::UTC );
+      metadata.lastModified = lastModifiedUtc.toLocalTime();
+    }
+  }
+  return true;
+}
+
+
+static bool _projectsTableExists( QgsOracleConn *conn, const QString &owner )
```
[the remark this answers is not in the document]
Comment (rouault) [reply src/providers/oracle/qgsoracleprojectstorage.cpp:50]: The rule is at https://en.cppreference.com/w/cpp/language/identifiers . I imagine what would match is: "the identifiers that begin with an underscore are reserved in the global namespace. "  . That said the likelihood a compiler / standard library would use "_projectsTableExists" for their own purpose is low.   But something like "_id" would be more risky

File: src/providers/oracle/qgsoracleprojectstorage.cpp [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    QgsDebugMsg( QStringLiteral( "SQL: %1\nERROR: %2" )
+                 .arg( qry.lastQuery() )
+                 .arg( qry.lastError().text() ) );
+    return false;
+  }
+
+  qry.addBindValue( owner );
+  qry.addBindValue( "qgis_projects" );
+
+  if ( !qry.exec() )
```
[the remark this answers is not in the document]
Comment (troopa81) [reply src/providers/oracle/qgsoracleprojectstorage.cpp]: done

File: src/providers/oracle/qgsoracleprojectstorage.cpp [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+  if ( !projectUri.valid )
+    return lst;
+
+  QgsPoolOracleConn pconn( projectUri.connInfo.connectionInfo( false ) );
+  if ( !pconn.get() )
+    return lst;
+
+  if ( _projectsTableExists( pconn.get(), projectUri.owner ) )
+  {
+    QString sql( QStringLiteral( "SELECT name FROM %1.\"qgis_projects\"" ).arg( QgsOracleConn::quotedIdentifier( projectUri.owner ) ) );
```
[the remark this answers is not in the document]
Comment (troopa81) [reply src/providers/oracle/qgsoracleprojectstorage.cpp]: Note sure, what's the befenit ? I spare a function call that would return the exact same string

File: src/providers/oracle/qgsoracleprojectstorage.cpp [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+bool QgsOracleProjectStorage::readProject( const QString &uri, QIODevice *device, QgsReadWriteContext &context )
+{
+  QgsOracleProjectUri projectUri = decodeUri( uri );
+  if ( !projectUri.valid )
+  {
+    context.pushMessage( QObject::tr( "Invalid URI for Oracle provider: " ) + uri, Qgis::MessageLevel::Critical );
+    return false;
+  }
+
+  QgsPoolOracleConn pconn( projectUri.connInfo.connectionInfo( false ) );
```
[the remark this answers is not in the document]
Comment (troopa81) [reply src/providers/oracle/qgsoracleprojectstorage.cpp:103]: yes

**Comment by troopa81** [comment]
@elpaso Thanks for reviewing, I treat your comments. Sorry for the delay I was away from computer

**Comment by troopa81** [comment]
unrelated grass test failure

**Pull request merged**